### PR TITLE
[mariadb-connector-cpp] New port

### LINKED ISSUE #39062: [New Port Request] mariadb-connector-cpp
### Library name

mariadb-connector-cpp

### Library description

MariaDB Connector for C++

### Source repository URL

https://github.com/mariadb-corporation/mariadb-connector-cpp

### Project homepage (if different from the source repository)

https://mariadb.com/docs/server/connect/programming-languages/cpp/

### Anything else that is useful to know when adding (such as optional features the library may have that should be included)

I have found the same issue #27571, but it was unresolved.
It must copy a library source code [mariadb-connector-c](https://github.com/mariadb-corporation/mariadb-connector-c) to directory called `libmariadb/`, and then it can be built.
I have forked [vcpkg](https://github.com/quqiOnfree/vcpkg), but I don't know how to modify it correctly.

### PR DESCRIPTION
**Pull request by dg0yt**
Based on #39185. Fixes https://github.com/microsoft/vcpkg/issues/39062.

Attention: Due to the libmysql/libmariadb conflict and other CI baseline entries, vcpkg CI runs only for x86-windows.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says. 
      Upstream points at the LGPL-2.1 file. "-or-later" is based on inspecting some source files.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by dg0yt** [comment]
CC @quqiOnfree for testing.

**Comment by quqiOnfree** [comment]
```
x64-linux - Succeed
x64-windows - Failed
```
[Details](https://github.com/Hcolda/QingLiaoChatServer/pull/16)

File: ports/mariadb-connector-cpp/install.diff [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+@@ -714,13 +716,28 @@ ELSE()
+     #MESSAGE(STATUS "Configuring to generate PKG package")
+     #ADD_SUBDIRECTORY(osxinstall)
+   ENDIF()
++  if(BUILD_SHARED_LIBS)
+   INSTALL(TARGETS ${LIBRARY_NAME}
++          EXPORT unofficial-mariadb-connector-cpp
++          RUNTIME DESTINATION ${INSTALL_BINDIR}
+           LIBRARY DESTINATION ${INSTALL_LIBDIR}
++          ARCHIVE DESTINATION ${INSTALL_LIBDIR}
```
Comment (dg0yt) [inline ports/mariadb-connector-cpp/install.diff:32]: `LIBRARY` is enough to install the import lib.

But TIL: 
`ARCHIVE` must be installed to make cmake export `IMPORTED_IMPLIB`.

**Comment by quqiOnfree** [comment]
Now, the library can be installed successfully, but where is the `conncpp.hpp`?
`error C1083: Cannot open include file: 'conncpp.hpp': No such file or directory `
[Details](https://github.com/Hcolda/QingLiaoChatServer/actions/runs/10094422197)

**Comment by dg0yt** [comment]
> `error C1083: Cannot open include file: 'conncpp.hpp': No such file or directory `

The file is installed, but `INTERFACE_INCLUDE_DIRECTORIES` isn't set...
FTR the sample apps uses this form:
~~~c++
#include <mariadb/conncpp.hpp>
~~~

**Comment by quqiOnfree** [comment]
> > `error C1083: Cannot open include file: 'conncpp.hpp': No such file or directory `
> 
> The file is installed, but `INTERFACE_INCLUDE_DIRECTORIES` isn't set... FTR the sample app uses this form:
> 
> ```c++
> #include <mariadb/conncpp.hpp>
> ```

Thanks, and I will change the form in the future.

**Comment by quqiOnfree** [comment]
```
x64-windows - success
```
Because I'm on a business trip these days, testing on Linux has to be postponed.

**Comment by quqiOnfree** [comment]
The port usage tests with the following triplets:
```
x64-linux - success
x64-mingw-dynamic - success
```

**Review by WangWeiLin-MV — DISMISSED** [review]
The port usage tests pass with the following triplets:

* x64-linux
* x64-windows

**Comment by BillyONeal** [comment]
> Attention: Due to the libmysql/libmariadb conflict and other CI baseline entries, vcpkg CI runs only for x86-windows.

Specifically:

https://github.com/microsoft/vcpkg/blob/1dc5ee30eb1032221d29f281f4a94b73f06b4284/scripts/ci.baseline.txt#L581-L592

https://github.com/microsoft/vcpkg/blob/1dc5ee30eb1032221d29f281f4a94b73f06b4284/scripts/ci.baseline.txt#L605

This has been skipped for at least 5 years ( https://github.com/microsoft/vcpkg/commit/92dc30ff52fae0d7236b5b3dd7e6b2f0ff39ae6f ); apparently this thing is from the era when we tried to accept some forms of alternatives.

File: ports/mariadb-connector-cpp/fix-carray.diff [hunk 2]
```diff
@@ -0,0 +1,12 @@
+diff --git a/include/conncpp/CArray.hpp b/include/conncpp/CArray.hpp
+index f3e4634..e0f62eb 100644
+--- a/include/conncpp/CArray.hpp
++++ b/include/conncpp/CArray.hpp
+@@ -24,6 +24,7 @@
+ #include "buildconf.hpp"
+ #include <initializer_list>
+ #include <vector>
++#include <cstdint>
```
Comment (BillyONeal) [inline ports/mariadb-connector-cpp/fix-carray.diff:9]:
```suggestion
+#include <stdint.h>
```

Upstream appears to be using the non `std::` ones:

https://github.com/mariadb-corporation/mariadb-connector-cpp/blob/b09555de99ed4b1d054a88ff85acbae996bce1d1/include/conncpp/CArray.hpp#L51

**Comment by BillyONeal** [comment]
I just want to double check with maintainers that we should take this given the mysql situation, I think the answer is probably yes.

File: ports/mariadb-connector-cpp/fix-carray.diff (hunk 2 above)
Comment (dg0yt) [reply ports/mariadb-connector-cpp/fix-carray.diff:9]: In fact I assume `cstdint` means `C` stdint.
I see no connection to nameespace `std::` on https://en.cppreference.com/w/cpp/header/cstdint but "This header was originally in the C standard library as [<stdint.h>](https://en.cppreference.com/w/c/types/integer)."

Comment (dg0yt) [reply ports/mariadb-connector-cpp/fix-carray.diff:9]: Ah I see the synopsis now.

**Comment by BillyONeal** [comment]
We agree that this is consistent with the status quo. Thanks for the new port!

**Pull request merged**